File: client/app/src/main/java/com/rh/heji/ui/bill/Iteminfo/BillInfoPop.java
package com.rh.heji.ui.bill.Iteminfo;

import android.content.Context;
import android.view.View;
import android.widget.ImageView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.blankj.utilcode.util.TimeUtils;
import com.blankj.utilcode.util.ToastUtils;
import com.lxj.xpopup.XPopup;
import com.lxj.xpopup.core.BottomPopupView;
import com.lxj.xpopup.util.XPopupUtils;
import com.rh.heji.AppCache;
import com.rh.heji.BuildConfig;
import com.rh.heji.R;
import com.rh.heji.data.AppDatabase;
import com.rh.heji.data.converters.DateConverters;
import com.rh.heji.data.db.Bill;
import com.rh.heji.data.db.Constant;
import com.rh.heji.data.db.Image;
import com.rh.heji.databinding.PopBilliInfoBinding;
import com.rh.heji.ui.bill.img.ImageLoader;
import com.rh.heji.ui.user.JWTParse;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;
import java.util.stream.Collectors;

/**
 * Date: 2020/9/20
 * Author: 锅得铁
 * #
 */
public class BillInfoPop extends BottomPopupView {
    private Bill bill;
    PopBilliInfoBinding binding;
    private ImageAdapter imageAdapter;
    private PopClickListener popClickListener;

    public BillInfoPop(@NonNull Context context) {
        super(context);
    }

    public Bill getBill() {
        return bill;
    }

    public void setBill(Bill bill) {
        this.bill = bill;
        binding.tvMonney.setText(bill.getMoney().toString());
        binding.tvType.setText(bill.getCategory());
        binding.tvRecordTime.setText(TimeUtils.millis2String(bill.getCreateTime()));
        binding.tvTicketTime.setText(DateConverters.date2Str(bill.getBillTime()));
        binding.rePeople.setText(bill.getDealer());
    }

    public void setBillImages(List<Image> images) {
        if (images.isEmpty()) {
            if (imageAdapter != null) {
                imageAdapter.setNewInstance(new ArrayList<>());
                imageAdapter.notifyDataSetChanged();
            }
            return;
        }

        initTicketImg();
        //服务器返回的是图片的ID、需要加上前缀
        List<Image> imagePaths = images.stream().map(image -> {
            String onlinePath = image.getOnlinePath();
            if (null != onlinePath && !image.getOnlinePath().contains("http")) {
                String path = BuildConfig.HTTP_URL + "/image/" + image.getOnlinePath();
                image.setOnlinePath(path);
            }
            return image;
        }).collect(Collectors.toList());
        imageAdapter.setNewInstance(imagePaths);
    }

    @Override
    protected int getImplLayoutId() {
        return R.layout.pop_billi_info;
    }

    @Override
    protected void onCreate() {
        super.onCreate();
        binding = PopBilliInfoBinding.bind(getPopupContentView().findViewById(R.id.billInfoCard));
        binding.tvDelete.setOnClickListener(v -> {
            deleteBill();
            if (popClickListener != null) {
                popClickListener.delete(bill.getId());
            }
        });
        binding.tvUpdate.setOnClickListener(v -> {
            if (popClickListener != null) {
                popClickListener.update(bill.getId());
            }
        });
        //设置圆角背景
        getPopupImplView().setBackground(XPopupUtils.createDrawable(getResources().getColor(R.color._xpopup_light_color),
                popupInfo.borderRadius, popupInfo.borderRadius, 0, 0));
    }

    private void initTicketImg() {
        binding.ticketRecycler.setLayoutManager(new LinearLayoutManager(getContext(), LinearLayoutManager.HORIZONTAL, false));

        imageAdapter = new ImageAdapter();
        binding.ticketRecycler.setAdapter(imageAdapter);

        imageAdapter.setDiffCallback(new DiffUtil.ItemCallback<Image>() {
            @Override
            public boolean areItemsTheSame(@NonNull Image oldItem, @NonNull Image newItem) {
                return oldItem.equals(newItem);
            }

            @Override
            public boolean areContentsTheSame(@NonNull Image oldItem, @NonNull Image newItem) {
                return oldItem.equals(newItem);
            }
        });
        imageAdapter.setOnItemClickListener((adapter, view, position) -> {
            showTicketImg(view, position);
        });

    }


    private void showTicketImg(View itemView, int itemPosition) {
        List<Object> objects = imageAdapter.getData().stream().map((Function<Image, Object>) image -> imageAdapter.getImagePath(image)).collect(Collectors.toList());
        if (objects.isEmpty()) return;
        new XPopup.Builder(getContext()).asImageViewer((ImageView) itemView, itemPosition, objects,
                (popupView, position1) -> {
                    popupView.updateSrcView((ImageView) imageAdapter.getViewByPosition(itemPosition, R.id.itemImage));
                }, new ImageLoader())
                .isShowSaveButton(false)
                .show();
    }

    /**
     * 删除该条账单
     */
    private void deleteBill() {
        new XPopup.Builder(getContext()).asConfirm("删除提示", "确认删除该条账单吗？",
                () -> {
                    JWTParse.User user = JWTParse.INSTANCE.getUser(AppCache.Companion.getInstance().getToken());
                    String createUser = bill.getCreateUser();
                    if (createUser == null || bill.getCreateUser().equals(user.getUsername())) {
                        bill.setSynced(Constant.STATUS_DELETE);
                        AppDatabase.getInstance().billDao().update(bill);
                        AppCache.Companion.getInstance().getAppViewModule().billDelete(bill.getId());
                        if (popClickListener != null) {
                            popClickListener.delete(bill.getId());
                        }
                    } else {
                        ToastUtils.showLong("只有账单创建人有权删除该账单");
                    }
                    BillInfoPop.this.dismiss();

                }).show();

    }

    public interface PopClickListener {
        void delete(String _id);

        void update(String _id);
    }

    public void setPopClickListener(PopClickListener popClickListener) {
        this.popClickListener = popClickListener;
    }
}

File: client/app/src/main/java/com/rh/heji/ui/bill/iteminfo/BillInfoPop.kt
package com.rh.heji.ui.bill.iteminfo

import android.view.View
import android.widget.ImageView
import androidx.lifecycle.Observer
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import com.blankj.utilcode.util.LogUtils
import com.blankj.utilcode.util.TimeUtils
import com.blankj.utilcode.util.ToastUtils
import com.chad.library.adapter.base.BaseQuickAdapter
import com.lxj.xpopup.XPopup
import com.lxj.xpopup.core.BottomPopupView
import com.lxj.xpopup.core.ImageViewerPopupView
import com.lxj.xpopup.util.XPopupUtils
import com.rh.heji.*
import com.rh.heji.data.converters.DateConverters
import com.rh.heji.data.db.Bill
import com.rh.heji.data.db.Image
import com.rh.heji.databinding.PopBilliInfoBinding
import com.rh.heji.ui.bill.add.AddBillFragmentArgs
import com.rh.heji.ui.bill.img.ImageLoader
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * Date: 2020/9/20
 * Author: 锅得铁
 * #
 */
class BillInfoPop(
    val activity: MainActivity,
    val bill: Bill,
    var popClickListener: BillPopClickListenerImpl = BillPopClickListenerImpl(),
) : BottomPopupView(activity), Observer<List<Image>> {
    private val imageObservable by lazy { activity.mainViewModel.getBillImages(billId = bill.id) }

    lateinit var binding: PopBilliInfoBinding
    private var imageAdapter = ImageAdapter()

    override fun getImplLayoutId(): Int {
        return R.layout.pop_billi_info
    }

    override fun onCreate() {
        super.onCreate()
        binding = PopBilliInfoBinding.bind(popupContentView.findViewById(R.id.billInfoCard))
        binding.tvDelete.setOnClickListener {
            deleteTip()
        }
        binding.tvUpdate.setOnClickListener {
            popClickListener?.update(bill)
            val bundle = AddBillFragmentArgs.Builder(bill).build().toBundle()
            activity.navController.navigate(R.id.nav_bill_add, bundle)
            dismiss()
        }
        //设置圆角背景
        popupImplView.background = XPopupUtils.createDrawable(
            resources.getColor(R.color._xpopup_light_color, null),
            popupInfo.borderRadius, popupInfo.borderRadius, 0f, 0f
        )
        binding.apply {
            tvMonney.text = bill.money.toString()
            tvType.text = bill.category
            tvRecordTime.text = bill.createTime?.let { TimeUtils.millis2String(it) }
            tvTicketTime.text = DateConverters.date2Str(bill.billTime)
            rePeople.text = bill.dealer
        }
        initBillImageList()//初始化列表和适配器
        if (bill.imgCount > 0) {
            imageObservable.observeForever(this)
        }
    }

    private fun initBillImageList() {
        binding.ticketRecycler.apply {
            layoutManager = LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false)
            adapter = imageAdapter
        }
        imageAdapter.setOnItemClickListener { adapter: BaseQuickAdapter<*, *>?, view: View, position: Int ->
            showImage(
                view,
                position
            )
        }
    }

    private fun showImage(itemView: View, itemPosition: Int) {
        val objects = imageAdapter.data
            .map { image: Image -> imageAdapter.getImagePath(image) }
            .toList()
        if (objects.isEmpty()) return
        XPopup.Builder(context)
            .asImageViewer(
                itemView as ImageView,
                itemPosition,
                objects,
                { popupView: ImageViewerPopupView, position: Int ->
                    popupView.updateSrcView(
                        imageAdapter.getViewByPosition(itemPosition, R.id.itemImage) as ImageView?
                    )
                },
                ImageLoader()
            )
            .isShowSaveButton(false)
            .show()
    }

    /**
     * 删除该条账单
     */
    private fun deleteTip() {
        XPopup.Builder(context).asConfirm(
            "删除提示", "确认删除该条账单吗？"
        ) {
            val mainActivity = context as MainActivity
            mainActivity.lifecycleScope.launch(Dispatchers.IO) {
                withContext(Dispatchers.Main) {
                    if (bill.createUser == currentUser.username) {
                        popClickListener.delete(bill)
                        dismiss()
                    } else {
                        ToastUtils.showLong("只有账单创建人有权删除该账单")
                    }
                }
            }

        }.show()
    }

    override fun onChanged(images: List<Image>) {
        if (images.isEmpty()) return
        //服务器返回的是图片的ID、需要加上前缀
        val imagePaths = images.map { image: Image ->
            val onlinePath = image.onlinePath
            if (onlinePath != null && !image.onlinePath!!.contains("http")) { //在线Image路径
                val path = BuildConfig.HTTP_URL + "/image/" + image.onlinePath
                image.onlinePath = path
            }
            image
        }.toMutableList()
        imageAdapter.setNewInstance(imagePaths)
    }

    override fun onDismiss() {
        super.onDismiss()
        if (bill.imgCount > 0) {
            imageObservable.removeObserver(this)
        }
    }
}

/**
 * 不对外开放该接口，通过覆写BillDefPopClickListener实现
 */
private interface PopClickListener {
    fun delete(bill: Bill)
    fun update(bill: Bill)
}

open class BillPopClickListenerImpl : PopClickListener {

    override fun delete(bill: Bill) {
        AppViewModule.get().billDelete(bill)
    }

    override fun update(bill: Bill) {
        LogUtils.d(bill)
    }

}